### PR TITLE
[Merged by Bors] - Add demo data-warehouse-iceberg-trino-spark

### PR DESCRIPTION
**Pull request by sbernauer**
## Description

Needs a larger k8s cluster! I use IONOS k8s with 9x 4 cores (8 threads), 20GB ram and 30GB hdd disk
Maybe we can also offer a smaller variant later on.

Otherwise business as usual. From feature-branch run `stackablectl --additional-stacks-file stacks/stacks-v1.yaml --additional-releases-file releases.yaml --additional-demos-file demos/demos-v1.yaml demo install data-warehouse-iceberg-trino-spark`

I'm not happy with some parts but i think an iterative approach is best:
* Shared bikes are currently not streamed into Kafka (instead on-time job)
* Some high-volume real-time datasource would be great. Currently we use the water levels and duplicate them to get higher volumes.
* Some sort of Upsert or Deletion usecases would be great. But probably not on the large datasets as for our wallet ^^
* Better Dashboards. The current one were thrown together quickly
* I would like to partitions the water_level measurements by day but run into https://github.com/apache/iceberg/issues/5625. There might be ways around by using a dedicated Spark context for compaction but we can easily adopt the partitioning after the issue gets fixed. Sorting during rewrites did cost performance during compaction and did not provide real benefits for my intial measurements. Disabled for now.
* As always tracked my findings in https://github.com/stackabletech/stackablectl/issues/128

To get to the Spark UI `kubectl port-forward $(kubectl get pod -o name | grep 'spark-ingest-into-warehouse-.*-driver') 4040`

## Review Checklist
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Review by maltesander — APPROVED** [review]
Nice. LGTM!

**Comment by sbernauer** [comment]
I kept running over the weekend. We now are a
```
|files|total_size     |avg_size   |total_records |avg_records|smallest_file|
|-----|---------------|-----------|--------------|-----------|-------------|
|298  |150,318,197,593|504,423,481|36,430,087,583|122,248,616|7,101,325    |
```

**Comment by sbernauer** [comment]
@maltesander could you please do another round reviewing the docs?

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 1]
```diff
@@ -0,0 +1,275 @@
+= data-warehouse-iceberg-trino-spark
+
+[WARNING]
+====
+This demos uses significant amount of resources. It will most likely not run on your workstation.
+It was developed and tested on 10 nodes with 4 cores (8 threads), 20GB RAM and 30GB hdd disks.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:6]:
```suggestion
It was developed and tested on 10 nodes with each 4 cores (8 threads), 20GB RAM and 30GB HDD.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+Additionally persistent volumes with a total size of approx. 1TB will get created.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:7]:
```suggestion
Additionally, persistent volumes with a total size of approximately 1TB will get created.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+A smaller version of this demo might be created in the future.
+====
+
+[NOTE]
+====
+This guide assumes you already have the demo `data-warehouse-iceberg-trino-spark` installed.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:13]:
```suggestion
This guide assumes that you already have the demo `data-warehouse-iceberg-trino-spark` installed.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+This demo will
+
+* Install the required Stackable operators
+* Spin up the following data products
+** *Trino*: A fast distributed SQL query engine for big data analytics that helps you explore your data universe. This demo uses it to enable SQL access to the data
+** *Spark*: A multi-language engine for executing data engineering, data science, and machine learning. This demo uses it to stream data from Kafka into the warehouse
+** *MinIO*: A S3 compatible object store. This demo uses it as persistent storage to store all the data used
+** *Kafka*:  A distributed event streaming platform for high-performance data pipelines, streaming analytics and data integration. This demos uses it as an event streaming platform to stream the data in near real-time
+** *NiFi*:  An easy-to-use, powerful system to process and distribute data. This demos uses it to fetch multiple online real-time datasources and ingest it into Kafka
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:26]:
```suggestion
** *Kafka*: A distributed event streaming platform for high-performance data pipelines, streaming analytics and data integration. This demos uses it as an event streaming platform to stream the data in near real-time
** *NiFi*: An easy-to-use, powerful system to process and distribute data. This demos uses it to fetch multiple online real-time data sources and ingest it into Kafka
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+** *Hive metastore*: A service that stores metadata related to Apache Hive and other services. This demo uses it as metadata storage for Trino and Spark
+** *Open policy agent* (OPA): A open source, general-purpose policy engine that unifies policy enforcement across the stack. This demo uses it as the authorizer for Trino, which decides which user is able to query which data.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:28]:
```suggestion
** *Open policy agent* (OPA): An open source, general-purpose policy engine that unifies policy enforcement across the stack. This demo uses it as the authorizer for Trino, which decides which user is able to query which data.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+** *Superset*: A modern data exploration and visualization platform. This demo utilizes Superset to retrieve data from Trino via SQL queries and build dashboards on top of that data
+* Copy multiple datasources in csv and parquet format into the s3 staging area
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:30]:
```suggestion
* Copy multiple data sources in CSV and Parquet format into the S3 staging area
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+* Let Trino copy the data from staging area into the warehouse area. During the copy transformations such as validating, casting, parsing timestamps and enriching the data by joining lookup-tables are done
+* Simultaneously start a NiFi workflow, which fetches real-time datasets via the internet and ingests the data as json records into Kafka
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:32]:
```suggestion
* Simultaneously start a NiFi workflow, which fetches datasets in real-time via the internet and ingests the data as JSON records into Kafka
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+You can see the deployed products as well as their relationship in the following diagram:
+
+image::demo-data-warehouse-iceberg-trino-spark/overview.png[]
+
+== Apache iceberg
+As Apache iceberg states on their https://iceberg.apache.org/docs/latest/[website]:
+
+> Apache Iceberg is an open table format for huge analytic datasets. Iceberg adds tables to compute engines including Spark, Trino, PrestoDB, Flink, Hive and Impala using a high-performance table format that works just like a SQL table.
+
+This demos uses Iceberg as it plays along nicely with Object storages as well as having a Trino and Spark integration.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:45]:
```suggestion
This demos uses Iceberg as it plays along nicely with object storage as well as having a Trino and Spark integration.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+The S3 provided by MinIO is used as persistent storage to store all the data used.
+Open the `minio` endpoint `console-http` retrieved by `stackablectl services list` in your browser (http://217.160.222.211:31429 in this case).
+
+image::demo-data-warehouse-iceberg-trino-spark/minio_1.png[]
+
+Log in with the username `admin` and password `adminadmin`.
+
+image::demo-data-warehouse-iceberg-trino-spark/minio_2.png[]
+
+Here you can see the two buckets the S3 is split into:
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:97]:
```suggestion
Here you can see the two buckets contained in the S3:
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+1. `staging`: The demo loads static datasets into this area. It is stored in different formats, such as csv and parquet. It does contain actual data tables as well as lookup tables.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:99]:
```suggestion
1. `staging`: The demo loads static datasets into this area. It is stored in different formats, such as CSV and Parquet. It does contain actual data tables as well as lookup tables.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+Go ahead and click on any folder.
+
+image::demo-data-warehouse-iceberg-trino-spark/minio_5.png[]
+
+You can see that Trino has placed a single file here containing all the house sales of that particular year.
+
+== NiFi
+
+NiFi is used to fetch multiple datasources from the internet and ingest it into Kafka near-realtime.
+Some of the datasources are statically download (e.g. as csv) and others are fetch via APIs such as REST APIs.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:123]:
```suggestion
Some data sources are statically downloaded (e.g. as CSV) and others are dynamically fetched via APIs such as REST APIs.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+This includes the following datasources:
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:124]:
```suggestion
This includes the following data sources:
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+You can have a look at the ingestion job running in NiFi by opening the given `nifi` endpoint `https` from your `stackablectl services list` command output (https://217.160.120.117:31499 in this case).
+If you get a warning regarding the self-signed certificate generated by the xref:secret-operator::index.adoc[Secret Operator] (e.g. `Warning: Potential Security Risk Ahead`), you have to tell your browser to trust the website and continue.
+
+image::demo-data-warehouse-iceberg-trino-spark/nifi_1.png[]
+
+Log in with the username `admin` and password `adminadmin`.
+
+image::demo-data-warehouse-iceberg-trino-spark/nifi_2.png[]
+
+As you can see, the NiFi workflow consists of lot's of components.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:143]:
```suggestion
As you can see, the NiFi workflow consists of lots of components.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+ zookeeper  zookeeper     default    zk                  212.227.224.138:32321                                                  
+                                                                                                                                
+ minio      minio         default    http                http://217.160.222.211:32031   Third party service                     
+                                     console-http        http://217.160.222.211:31429   Admin user: admin, password: adminadmin 
+----
+
+[NOTE]
+====
+When a product instance has not finished starting yet, the service will have no endpoint.
+Starting all of the product instances might take a considerable amount of time depending on your internet connectivity.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:82]:
```suggestion
Starting all the product instances might take a considerable amount of time depending on your internet connectivity.
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+Log in with the username `admin` and password `adminadmin`.
+
+image::demo-data-warehouse-iceberg-trino-spark/nifi_2.png[]
+
+As you can see, the NiFi workflow consists of lot's of components.
+You can zoom in by using your mouse and mouse wheel.
+On the left side are two strands, that
+
+1. Fetch the list of known water-level stations and ingests them into Kafka
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:147]:
```suggestion
1. Fetch the list of known water-level stations and ingest them into Kafka
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+2. Continuously run a loop fetching the measurements of the last 30 for every measuring stations and ingesting the measurements into Kafka
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:148]:
```suggestion
2. Continuously run a loop fetching the measurements of the last 30 for every measuring station and ingesting the measurements into Kafka
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+
+On the right side are three strands, that
+
+1. Fetch the current shred bike stations information
+2. Fetch the current shred bike stations status
+3. Fetch the current shred bike bike status
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:154]: bike bike status :) ?
```suggestion
1. Fetch the current shared bike stations information
2. Fetch the current shared bike stations status
3. Fetch the current shared bike bike status
```

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+
+=== Connect with DBeaver
+https://dbeaver.io/[DBeaver] is free multi-platform database tool that can be used to connect to Trino.
+Please have a look at the <TODO> trino-operator documentation on how to connect DBeaver to Trino.
+
+image::demo-data-warehouse-iceberg-trino-spark/dbeaver_1.png[]
+
+image::demo-data-warehouse-iceberg-trino-spark/dbeaver_2.png[]
+You need to modify the setting `TLS` to `true`.
+Additionally no need to add the setting `SSLVerification` and set it to `NONE`.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:206]:
```suggestion
Additionally, set `SSLVerification` to `NONE`.
```

**Review by maltesander — CHANGES_REQUESTED** [review]
Just english nitpicking!

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc (hunk 18 above)
Comment (sbernauer) [reply docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:206]: This actually is a 2-step process. First adding, later setting the value

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+
+image::demo-data-warehouse-iceberg-trino-spark/trino_1.png[]
+
+Log in with the username `admin` and password `admin`.
+
+image::demo-data-warehouse-iceberg-trino-spark/trino_2.png[]
+
+=== Connect with DBeaver
+https://dbeaver.io/[DBeaver] is free multi-platform database tool that can be used to connect to Trino.
+Please have a look at the <TODO> trino-operator documentation on how to connect DBeaver to Trino.
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:200]: Extracted into https://github.com/stackabletech/trino-operator/issues/321

[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:200]: We should create a ticket for that once its done to adapt it here?

[the remark this answers is not in the document]
Comment (sbernauer) [reply docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:200]: Its in the definition of done in that ticket ;)

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc (hunk 18 above)
Comment (sbernauer) [reply docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:206]: Arghhhh, nevermind. Setting it to

Additionally, you need to add the setting `SSLVerification` and set it to `NONE`

File: docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc (hunk 19 above)
[the remark this answers is not in the document]
Comment (maltesander) [reply docs/modules/ROOT/pages/demos/data-warehouse-iceberg-trino-spark.adoc:200]: OK

**Review by maltesander — APPROVED** [review]
LGTM!

**Comment by sbernauer** [comment]
Many many thanks!

**Pull request closed without merging**